rules: Make ADB and fastboot work out-of-the-box

luk1337 opened the pull request:
https://android.googlesource.com/platform/packages/modules/adb/+/d0db47dcdf941673f405e1095e6ffb5e565902e5/adb.h#199
https://android.googlesource.com/platform/system/core/+/7199051aaf0ddfa2849650933119307327d8669c/fastboot/fastboot.cpp#244

yuwata commented on rules.d/70-uaccess.rules.in:
```diff
@@ -77,6 +77,10 @@ ENV{DDC_DEVICE}=="?*", TAG+="uaccess"
 
+# Android devices (ADB DBC / ADB)
+SUBSYSTEM=="usb", ENV{ID_USB_INTERFACES}=="*:dc02??:*", TAG+="uaccess"
+SUBSYSTEM=="usb", ENV{ID_USB_INTERFACES}=="*:ff42??:*", TAG+="uaccess"
```
I do not think we should add such rules that directly matching with specific ID_USB_INTERFACE anymore.
Instead, please introduce a short hwdb file, which sets e.g. ID_ANDROID_DEBUG_BRIDGE=1, and make the rule check the property.

See e.g. 70-av-production.hwdb which sets ID_AV_XYZ, and related rules below.

luk1337 replied:
I'm not sure how to match the USB interfaces in hwdb, I tried this:

```
usb:v*p*d*dc*dsc*dp*icFFisc42*
 ID_ANDROID_DEBUG_BRIDGE=1
```

```
SUBSYSTEM=="usb", ENV{ID_ANDROID_DEBUG_BRIDGE}=="1", TAG+="uaccess"
```

and while `udevadm test /devices/pci0000:00/0000:00:01.3/0000:03:00.2/0000:20:04.0/0000:25:00.0/usb3/3-1/3-1:1.0` shows ID_ANDROID_DEBUG_BRIDGE=1, `udevadm test /devices/pci0000:00/0000:00:01.3/0000:03:00.2/0000:20:04.0/0000:25:00.0/usb3/3-1` doesn't and thus adb devices can't access the device.

luk1337 replied:
Also, I don't want to match by vid/pid because there are a lot of them (see https://github.com/M0Rf30/android-udev-rules/blob/main/51-android.rules)

poettering replied:
I think matching against generic interfaces in udev rules is actually OK. matching against vid/pid should happen in hwdb if needed. 70-av-production.hwdb contains vid/pid matches only, too.

poettering replied:
are these interface ids actually documented anywhere?

luk1337 replied:
see link in my commit msg (https://android.googlesource.com/platform/packages/modules/adb/+/refs/heads/android15-release/adb.h#199)

poettering replied:
@M0Rf30 any chance you have some input on this?

poettering replied:
bInterfaceClass being "vendor specific" suggests that any other vendor might use the same class id for totally different purposes, no?

luk1337 replied:
and lsusb from fastboot

```
Bus 003 Device 003: ID 0fce:0dde Sony Ericsson Mobile Communications AB Android
Negotiated speed: High Speed (480Mbps)
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.10
  bDeviceClass            0 [unknown]
  bDeviceSubClass         0 [unknown]
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x0fce Sony Ericsson Mobile Communications AB
  idProduct          0x0dde Android
  bcdDevice            1.00
  iManufacturer           1 Google
  iProduct                2 Android
  iSerial                 3 HQ00000000
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0020
    bNumInterfaces          1
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0x80
      (Bus Powered)
    MaxPower              100mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass       255 Vendor Specific Class
      bInterfaceSubClass     66 [unknown]
      bInterfaceProtocol      3 
      iInterface              4 fastboot
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0200  1x 512 bytes
        bInterval               0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x01  EP 1 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0200  1x 512 bytes
        bInterval               1
Binary Object Store Descriptor:
  bLength                 5
  bDescriptorType        15
  wTotalLength       0x002a
  bNumDeviceCaps          3
  USB 2.0 Extension Device Capability:
    bLength                 7
    bDescriptorType        16
    bDevCapabilityType      2
    bmAttributes   0x00000006
      BESL Link Power Management (LPM) Supported
  SuperSpeed USB Device Capability:
    bLength                10
    bDescriptorType        16
    bDevCapabilityType      3
    bmAttributes         0x00
    wSpeedsSupported   0x000e
      Device can operate at Full Speed (12Mbps)
      Device can operate at High Speed (480Mbps)
      Device can operate at SuperSpeed (5Gbps)
    bFunctionalitySupport   1
      Lowest fully-functional device speed is Full Speed (12Mbps)
    bU1DevExitLat           7 micro seconds
    bU2DevExitLat         101 micro seconds
  SuperSpeedPlus USB Device Capability:
    bLength                20
    bDescriptorType        16
    bDevCapabilityType     10
    bmAttributes         0x00000001
      Sublink Speed Attribute count 2
      Sublink Speed ID count 1
    wFunctionalitySupport   0x1100
      Min functional Speed Attribute ID: 0
      Min functional RX lanes: 1
      Min functional TX lanes: 1
    bmSublinkSpeedAttr[0]   0x000a4030
      Speed Attribute ID: 0 10Gb/s Symmetric RX SuperSpeedPlus
    bmSublinkSpeedAttr[1]   0x000a40b0
      Speed Attribute ID: 0 10Gb/s Symmetric TX SuperSpeedPlus
Device Status:     0x0000
  (Bus Powered)
```

luk1337 replied:
> bInterfaceClass being "vendor specific" suggests that any other vendor might use the same class id for totally different purposes, no?

I guess? But at worst, we just tag it as uaccess?

poettering replied:
well, i am not sure we should open up potentially sensitive devices willy-nilly like that, if we have no clear indicator what it is we are talking to.

zenczykowski replied:
I think the chance of accidental collision is very low, and that the convenience of having adb/fastboot and webusb based android os flashers just work wins out...

poettering commented:
See: #23468

zenczykowski commented:
It looks like you can match:
      bInterfaceClass       255 Vendor Specific Class
      bInterfaceSubClass     66 [unknown]
      bInterfaceProtocol      1 
which is good, but could randomly (though I think super unlikely) conflict with some other devices.
(since technically this is vendor specific / unallocated / undefined space)

Ideally we'd also be matching one or both of:
    iConfiguration          4 adb
      iInterface              5 ADB Interface
which AFAICT has an actual text string description...

But yeah, I'm not sure if those are actually guaranteed to be there...

And the following only checks the first 3 ints:

https://cs.android.com/android/platform/superproject/main/+/main:packages/modules/adb/client/transport_usb.cpp;drc=4e2f31db1e8539352e42c1b523204242b97018e9;l=170

So...

luk1337 commented:
> > It looks like you can match: bInterfaceClass 255 Vendor Specific Class bInterfaceSubClass 66 [unknown] bInterfaceProtocol 1 which is good, but could randomly (though I think super unlikely) conflict with some other devices. (since technically this is vendor specific / unallocated / undefined space)
> > Ideally we'd also be matching one or both of: iConfiguration 4 adb iInterface 5 ADB Interface which AFAICT has an actual text string description...
> > But yeah, I'm not sure if those are actually guaranteed to be there...
> > And the following only checks the first 3 ints:
> > https://cs.android.com/android/platform/superproject/main/+/main:packages/modules/adb/client/transport_usb.cpp;drc=4e2f31db1e8539352e42c1b523204242b97018e9;l=170
> > So...
> 
> I skipped bInterfaceProtocol on purpose, as otherwise I'd have to add extra rules for fastboot.

But if we want to do that, that'd likely work:

```
# ADB DbC
SUBSYSTEM=="usb", ENV{ID_USB_INTERFACES}=="*:dc0201:*", TAG+="uaccess"

# ADB
SUBSYSTEM=="usb", ENV{ID_USB_INTERFACES}=="*:ff4201:*", TAG+="uaccess"

# Fastboot
SUBSYSTEM=="usb", ENV{ID_USB_INTERFACES}=="*:ff4203:*", TAG+="uaccess"
```

zenczykowski commented:
It's probably best to be as specific as possible to limit the chance of a random non-adb device matching.

luk1337 commented:
> It's probably best to be as specific as possible to limit the chance of a random non-adb device matching.

ok, I updated my change and added reference URL for fastboot interface (https://android.googlesource.com/platform/system/core/+/refs/heads/android15-release/fastboot/fastboot.cpp#244)

yuwata commented:
Could you update the links to make them reference a specific commit rather than HEAD ?
Otherwise, if the files are updated, it may become hard to find the referenced code.

luk1337 commented:
> Could you update the links to make them reference a specific commit rather than HEAD ? Otherwise, if the files are updated, it may become hard to find the referenced code.

Done.

yuwata reviewed (approved):
We already have similar rules for PTP/MTP protocol devices. If it is not possible or hard to do support ADB devices with hwdb, I think it is OK to merge this in the current form.

Leave @poettering for final approval.

poettering commented:
I'd really like some input from @M0Rf30, I think he might know a thing or two about this.

The google adb code what does it actually match against precisely when opening up a connection to usb devices?

poettering commented:
i am not sure i am onboard with matching against such vague identifiers generically. but if we do end up merging this, it needs an explanation in the udev rules files that explains the situation and the implications.

luk1337 commented:
> I'd really like some input from @M0Rf30, I think he might know a thing or two about this.
> 
> The google adb code what does it actually match against precisely when opening up a connection to usb devices?

https://android.googlesource.com/platform/packages/modules/adb/+/d0db47dcdf941673f405e1095e6ffb5e565902e5/adb.h#199
https://android.googlesource.com/platform/packages/modules/adb/+/d0db47dcdf941673f405e1095e6ffb5e565902e5/client/transport_usb.cpp#204
https://android.googlesource.com/platform/packages/modules/adb/+/d0db47dcdf941673f405e1095e6ffb5e565902e5/client/usb_linux.cpp#220

luk1337 commented:
> i am not sure i am onboard with matching against such vague identifiers generically. but if we do end up merging this, it needs an explanation in the udev rules files that explains the situation and the implications.

If someone would like to help with writing such disclaimer, feel free to suggest one, and I'll just update my PR with it.

poettering commented:
> https://android.googlesource.com/platform/packages/modules/adb/+/d0db47dcdf941673f405e1095e6ffb5e565902e5/adb.h#199
> https://android.googlesource.com/platform/packages/modules/adb/+/d0db47dcdf941673f405e1095e6ffb5e565902e5/client/transport_usb.cpp#204
> https://android.googlesource.com/platform/packages/modules/adb/+/d0db47dcdf941673f405e1095e6ffb5e565902e5/client/usb_linux.cpp#220

the latter definitely checks one more thing: the number of interfaces... i wonder if that's load bearing.

luk1337 commented:
> > https://android.googlesource.com/platform/packages/modules/adb/+/d0db47dcdf941673f405e1095e6ffb5e565902e5/adb.h#199
> > https://android.googlesource.com/platform/packages/modules/adb/+/d0db47dcdf941673f405e1095e6ffb5e565902e5/client/transport_usb.cpp#204
> > https://android.googlesource.com/platform/packages/modules/adb/+/d0db47dcdf941673f405e1095e6ffb5e565902e5/client/usb_linux.cpp#220
> 
> the latter definitely checks one more thing: the number of interfaces... i wonder if that's load bearing.

It checks number of endpoints, not interfaces. I assume that's just a sanity check to ensure there's input and output endpoint.

zenczykowski commented:
It certainly looks like a sanity check, since later there's comments about figuring out which one is IN vs OUT.  It does need a bidirectional pipe to function - and thus two endpoints (IN + OUT).

But this of course does imply in practice it does need to have exactly two...

luk1337 commented:
> > i am not sure i am onboard with matching against such vague identifiers generically. but if we do end up merging this, it needs an explanation in the udev rules files that explains the situation and the implications.
> 
> If someone would like to help with writing such disclaimer, feel free to suggest one, and I'll just update my PR with it.

```
# Android devices (ADB DbC, ADB, Fastboot)
# 
# The bInterfaceClass and bInterfaceSubClass used are documented in source code here:
# * https://android.googlesource.com/platform/packages/modules/adb/+/d0db47dcdf941673f405e1095e6ffb5e565902e5/adb.h#199
# * https://android.googlesource.com/platform/system/core/+/7199051aaf0ddfa2849650933119307327d8669c/fastboot/fastboot.cpp#244
#
# Since it's using a generic vendor specific interface class, this can potentially
# result in a rare case where non-ADB/Fastboot device ends up with an uaccess tag.
SUBSYSTEM=="usb", ENV{ID_USB_INTERFACES}=="*:dc0201:*", TAG+="uaccess"
SUBSYSTEM=="usb", ENV{ID_USB_INTERFACES}=="*:ff4201:*", TAG+="uaccess"
SUBSYSTEM=="usb", ENV{ID_USB_INTERFACES}=="*:ff4203:*", TAG+="uaccess"
```

does that sound good?

zenczykowski commented:
I updated (added comments) at:

https://usb-ids.gowdy.us/read/UC/dc/02/01

https://usb-ids.gowdy.us/read/UC/dc/02/03 (extrapolated)

https://usb-ids.gowdy.us/read/UC/ff/42/01

https://usb-ids.gowdy.us/read/UC/ff/42/03

poettering commented:
> SUBSYSTEM=="usb", ENV{ID_USB_INTERFACES}=="*:dc0201:*", TAG+="uaccess"
> SUBSYSTEM=="usb", ENV{ID_USB_INTERFACES}=="*:ff4201:*", TAG+="uaccess"
> SUBSYSTEM=="usb", ENV{ID_USB_INTERFACES}=="*:ff4203:*", TAG+="uaccess"

please merge into:

```
SUBSYSTEM=="usb", ENV{ID_USB_INTERFACES}=="*:dc0201:*|*:ff4201:*|*:ff4203:*", TAG+="uaccess"
```

poettering commented:
> does that sound good?

comment should also maybe briefly mention what these three interfaces actually are

luk1337 commented:
> > does that sound good?
> 
> comment should also maybe briefly mention what these three interfaces actually are

```
# Android devices (ADB DbC, ADB, Fastboot)
# Used to interact with devices over Android Debug Bridge and Fastboot protocols, see:
# * https://developer.android.com/tools/adb
# * https://source.android.com/docs/setup/test/running
# * https://source.android.com/docs/setup/test/flash
#
# The bInterfaceClass and bInterfaceSubClass used are documented in source code here:
# * https://android.googlesource.com/platform/packages/modules/adb/+/d0db47dcdf941673f405e1095e6ffb5e565902e5/adb.h#199
# * https://android.googlesource.com/platform/system/core/+/7199051aaf0ddfa2849650933119307327d8669c/fastboot/fastboot.cpp#244
#
# Since it's using a generic vendor specific interface class, this can potentially
# result in a rare case where non-ADB/Fastboot device ends up with a uaccess tag.
SUBSYSTEM=="usb", ENV{ID_USB_INTERFACES}=="*:dc0201:*|*:ff4201:*|*:ff4203:*", TAG+="uaccess"
```

that's ok?

luk1337 commented:
> hmm, thinking about this I'd prefer one level of indirection here via a property. i.e match against the interfaces to set ID_APPLIANCE_DEBUG=1 or so (the name is just an idea, pick something else if you can come with something better, that is generic enough), and then match against that to enable uaccess. i.e. we I think devices should be identifiable as what they are, potentially other matches even via hwdb should be able to set the prop too, and we want some level of debuggability in case people wonder why uaccess was enabled.

you can see my failed attempt at matching devices by interface class here https://github.com/systemd/systemd/pull/36939#discussion_r2023731800

poettering commented:
> you can see my failed attempt at matching devices by interface class in hwdb here [#36939 (comment)](https://github.com/systemd/systemd/pull/36939#discussion_r2023731800)

no, you got me wrong. hwdb should be used for matching vid/pid, not for matching interfaces.

we want to use hwdb to contain checks for specific implementations of hw via vid/pid, but generic matches against entire classes of devices (which the interface check is) are fine to be done with rules.

if the usb interface check you proposed here works, then this means we hopefully never have to populate the hwdb with additional vid/pid checks for any of this, but I still thinkg it makes sense to introduce that indirection property, so that if the nee arises people *could* mark their devices like that.

I.e. I am suggesting:

```
SUBSYSTEM=="usb", ENV{ID_USB_INTERFACES}=="*:dc0201:*|*:ff4201:*|*:ff4203:*", ENV{ID_DEBUG_APPLIANCE}="android"
ENV{ID_DEBUG_APPLIANCE}=="?*", TAG+="uaccess"
```

i.e. the first line marks android adb stuff generically as debug appliance stuff. and the second line then adds uaccess to that. and if other types of devices (e.g. iphone usb debug stuff if that exists) one day wants to be hooked up that way too they could just use ID_DEBUG_APPLIANCE to something non-empty, too.

zenczykowski commented:
Do we need any additional testing or anything to get this pull request merged?

luk1337 commented:
bump?

luk1337 commented:
bump2?

poettering commented:
Did you see the comments over on https://github.com/M0Rf30/android-udev-rules/issues/344?

poettering commented:
(no idea what to make of those comments though.)

poettering commented:
my take on it:

https://github.com/M0Rf30/android-udev-rules/issues/344#issuecomment-2811963881

luk1337 commented:
> Did you see the comments over on [M0Rf30/android-udev-rules#344](https://github.com/M0Rf30/android-udev-rules/issues/344)?

I just read them, but for me, it looks like most of these issues were because of matching by vendor id and changing group/mode instead of using uaccess tag? I can't imagine TAG+="uaccess" would prevent something from accessing the device, unless I just don't exactly understand what it does.

zenczykowski commented:
Made a comment over there as well, but in general I'm also not seeing the problem, afaict 'uaccess' just adds a/some extra extended acl(s) to the usb nodes.  Worst case a program *manually* looks at node uid/gid (which is unchanged) and does manual permission checks (and thus incorrectly ignores extended acls) and stuff fails, but that wouldn't be a regression, since it would already be failing.  Anything that just relies on the kernel should IMHO 'just' work...

luk1337 commented:
can we have this backported to v257 too?

yuwata commented:
> can we have this backported to v257 too?

We backport only bug fixes (and test improvements). Hence, we will not backport this.
